src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java

```diff
@@ -184,10 +184,17 @@ public int read(String name, ByteBuffer dst) throws IOException {
                 long address = nb.address();
                 int n = read(name, address, rem);
 
-                // copy from buffer into backing array
-                int off = dst.arrayOffset() + pos + Unsafe.ARRAY_BYTE_BASE_OFFSET;
-                unsafe.copyMemory(null, address, dst.array(), off, n);
-                dst.position(pos + n);
+                if (dst.hasArray()) {
+                    // copy from buffer into backing array
+                    int off = dst.arrayOffset() + pos + Unsafe.ARRAY_BYTE_BASE_OFFSET;
+                    unsafe.copyMemory(null, address, dst.array(), off, n);
+                    dst.position(pos + n);
+                } else {
+                    // backing array not accessible so transfer via temporary array
+                    byte[] tmp = new byte[n];
+                    unsafe.copyMemory(null, address, tmp, 0, n);
+                    dst.put(tmp);
+                }
```

Comment:
The changes look okay but I'm wondering what cases you have in mind so that we can add tests. We lack tests that use of these APIs with buffers that are created with JNI NewDirectByteBuffer.

Reply:
Tbh, I don't know any specific ByteBuffer implementation (other than DirectByteBuffer) that doesn't have an array, but in theory it is explicitly allowed. And since this `read(String name, ByteBuffer dst)` is public API, it should be prepared to deal with any allowed params.

But after all I didn't encounter any exception, I just copied the implementation from `write()` as I deemed it reasonable.

Alternatively, if ByteBuffers were sealed classes and only allow HeapByteBuffers (known to have an array) and DirectByteBuffers (allowing direct access to the address, thus irrelevant in this part of the code) we could skip the distinction entirely. But I guess that isn't something we can decide in the scope of this PR.

Regarding test cases: Other than creating a ByteBuffer mock that returns false upon `hasArray()`, I don't know how to verify this branch. That said, so far I failed to understand what framework or pattern to use to write JDK tests. The UserDefinedAttributeView test seems to be a mere `main(String[] args)` class that must not complete exceptionally.

Reply:
Buffers are closed abstraction and cannot be implemented outside of the java.nio package. When sealed classes are promoted from preview to be a permanent feature then maybe we should look at make them sealed. For now, I think the only scenario where the reported issue can arise is when someone injects their own class into java.nio, maybe a mock as you suggest. I'm not opposed to handling such cases but I suspect it would require work in several other areas of the JDK that currently have to distinguish between heap and direct buffers.

As regards tests, then yes, the basic unit test for UserDefinedAttributeView is an old style jtreg test. It pre-dates support for TestNG and other frameworks.

Reply:
If we can be sure that the buffers are either direct or have an array, I suggest we close this PR unresolved, change the write method and add a corresponding assertion to both methods.

Reply:
Closing this PR/issue and drop the code from write sounds good.

Reply:
Created a new issue to remove the code from `write(...)`: [JDK-8264341](https://bugs.openjdk.java.net/browse/JDK-8264341)